test/read.jl

```diff
@@ -0,0 +1,245 @@
+mktempdir() do dir
+
+tasks = []
+
+# Create test file...
+filename = joinpath(dir, "file.txt")
+text = "C1,C2\n1,2\na,b\n"
+
+# List of IO producers...
+l = Vector{Tuple{AbstractString,Function}}()
+
+
+# File
+io = (text) -> begin
+    open(io-> write(io, text), filename, "w")
+    Base.Filesystem.open(filename, Base.Filesystem.JL_O_RDONLY)
+end
+s = io(text)
+@test isa(s, IO)
+@test isa(s, Base.Filesystem.File)
+close(s)
+push!(l, ("File", io))
+
+
+# IOStream
+io = (text) -> begin
+    open(io-> write(io, text), filename, "w")
+    open(filename)
+end
+s = io(text)
+@test isa(s, IO)
+@test isa(s, IOStream)
+close(s)
+push!(l, ("IOStream", io))
+
+
+# IOBuffer
+io = (text)->IOBuffer(text)
+s = io(text)
+@test isa(s, IO)
+@test isa(s, IOBuffer)
+close(s)
+push!(l, ("IOBuffer", io))
+
+
+# TCPSocket
+
+# PR#14627
+Base.connect!(sock::TCPSocket, addr::Base.InetAddr) = Base.connect!(sock, addr.host, addr.port)
+
+addr = Base.InetAddr(ip"127.0.0.1", 4444)
+io = (text) -> begin
+    c = Condition()
+    tsk = @async begin
+        srv = listen(addr)
+        notify(c)
+        sock = accept(srv)
+        write(sock,text)
+        close(sock)
+        close(srv)
+    end
+    push!(tasks, tsk)
+    wait(c)
+    connect(addr)
+end
+s = io(text)
+@test isa(s, IO)
+@test isa(s, TCPSocket)
+close(s)
+push!(l, ("TCPSocket", io))
+
+
+@windows ? nothing : begin
+
+# PipeEndpoint
+socketname = joinpath(dir, "socket")
```

Comment:
this test should work on windows, but the socket pipes need to be in the right namespace (`"\\\\.\\pipe\\<socketname>"`, i think)